core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -1042,7 +1046,19 @@ private[spark] class TaskSetManager(
       // bound based on that.
       logDebug("Task length threshold for speculation: " + threshold)
       for (tid <- runningTasksSet) {
-        foundTasks |= checkAndSubmitSpeculatableTask(tid, time, threshold)
+        var speculated = checkAndSubmitSpeculatableTask(tid, time, threshold)
+        if (!speculated && tidToExecutorKillTimeMapping.contains(tid)) {
```

Comment:
I am just curious why this solution (introducing the kill time per task ID map: `tidToExecutorKillTimeMapping`) is chosen over storing the kill time per executor (`executorToKillTimeMapping` or something like that). In the later case here would be  something like: 

```scala
if (!speculated && executorToKillTimeMapping.nonEmpty) {
  val taskInfo = taskInfos(tid)
  executorToKillTimeMapping.get(taskInfo.executorId).foreach {  executorKillTime =>
  ...
  }
}
```

Reply:
Yes - We can store executorToKillTimeMapping also. In that case, we have to clean the map when executor is removed. For that we may have to make changes to pass "removeExecutor" signal till TaskSetManager.

I didn't see much difference in both the approaches as total entries we will be creating for a given executor in current approach will be equal to total tasks running for that executor (which shouldn't be a lot).